src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8672,11 +8586,37 @@ static bool tryResolveExtensionMethod(
                     CombineExtensionMethodArguments(left, analyzedArguments, actualArguments);
                 }
 
-                result = resolveOverloads(binder, methodGroup, actualArguments, options, returnType, returnRefKind, in callingConvention, expression, diagnostics);
-                diagnostics.Free();
-                return true;
+                var extensionMethodResult = resolveOverloads(this, methodGroup, actualArguments, options, returnType, returnRefKind, in callingConvention, expression, diagnostics);
+                // If the search in the current scope resulted in any applicable method (regardless of whether a best
+                // applicable method could be determined) then our search is complete. Otherwise, store aside the
+                // first non-applicable result and continue searching for an applicable result.
+                if (extensionMethodResult.HasAnyApplicableMethod)
+                {
+                    firstResult.Free();
+                    actualArguments?.Free();
+                    lookupResult.Free();
+                    classicExtensionLookupResult.Free();
+                    diagnostics.Free();
+                    return extensionMethodResult;
+                }
+                else if (firstResult.IsEmpty)
+                {
+                    firstResult = extensionMethodResult;
+                }
+                else
+                {
+                    // Neither the first result, nor applicable. No need to save result.
+                    extensionMethodResult.Free();
+                }
             }
 
+            Debug.Assert((actualArguments == null) || !firstResult.IsEmpty);
+            lookupResult.Free();
+            classicExtensionLookupResult.Free();
+            diagnostics.Free();
```

Comment:
>diagnostics.Free(); [](http://example.com/codeflow?start=12&length=19)

It looks like we might be dropping a non empty bag that wasn't captured anywhere.

Reply:
Yup, that's the current behavior in `main`. We only report the diagnostics up to the first result.
Let me know if you'd like me to leave a PROTOTYPE comment

Reply:
I do not see code in main doing this. I assume this location maps to [https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L8486](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L8486)